### PR TITLE
[wip] support node config modification

### PR DESCRIPTION
**Pull request by MonzElmasry**

### CODE REVIEW

File: Dockerfile.dapper [hunk 1]
```diff
@@ -0,0 +1,29 @@
+FROM golang:1.13.4-alpine3.10
```
Comment (briandowns) [inline Dockerfile.dapper:1]: Should this be in line with k3s? `FROM golang:1.15.5-alpine3.12`

File: package/Dockerfile [hunk 2]
```diff
@@ -27,6 +27,5 @@ ARG TAG
 RUN apk --no-cache add \
     jq libselinux-utils
 COPY --from=verify /opt/k3s /opt/k3s
-COPY scripts/upgrade.sh /bin/upgrade.sh
-ENTRYPOINT ["/bin/upgrade.sh"]
-CMD ["upgrade"]
+COPY bin/k3s-upgrade /usr/bin/
+ENTRYPOINT ["/usr/bin/k3s-upgrade"]
```
Comment (briandowns) [inline package/Dockerfile:31]:
```suggestion
ENTRYPOINT ["/usr/bin/k3s-upgrade"]

```

File: Dockerfile.dapper (hunk 1 above)
Comment (MonzElmasry) [reply Dockerfile.dapper]: fixed, Thank you

File: main.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+func main() {
+	app := cli.NewApp()
+	app.Name = "k3s-upgrade"
+	app.Commands = cli.Commands{
+		cli.Command{
+			Name: "prepare",
+			Flags: []cli.Flag{
+				cli.StringFlag{
+					Name:        "masterPlan",
```
Comment (galal-hussein) [inline main.go:30]: to be consistent, this can be master-plan

File: main.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+					EnvVar:      "masterplan",
```
Comment (galal-hussein) [inline main.go:31]: the env var should be capitalized

File: pkg/upgrade/upgrade.go [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+	return -1, fmt.Errorf("k3s process is not running")
+}
+
+func GetBinPath(pid int) (string, error) {
+	_, err := os.FindProcess(pid)
+	if err != nil {
+		return "", fmt.Errorf(" process is not found")
+	}
+	if pid == 1 {
+		return "/bin/k3s", nil
```
Comment (galal-hussein) [inline pkg/upgrade/upgrade.go:60]: this is for k3d I assume?

Comment (MonzElmasry) [reply pkg/upgrade/upgrade.go:60]: yes

File: pkg/upgrade/upgrade.go [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+	}
+	binPath := path.Join("/host/proc", strconv.Itoa(pid), "cmdline")
```
Comment (galal-hussein) [inline pkg/upgrade/upgrade.go:62]: so this is not the right bin path, the **content** of this file is actually the bin path

File: pkg/upgrade/upgrade.go [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+	proc, err := os.FindProcess(pid)
+	if err != nil {
+		fmt.Errorf(" process is not found")
+	}
+	return proc.Signal(syscall.SIGTERM)
+}
+
+func MatchSelector(hostPath, configPath string, preservedEntries []string) error {
+	kubeconfig := os.Getenv("KUBECONFIG")
+	conf, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
```
Comment (galal-hussein) [inline pkg/upgrade/upgrade.go:79]: so since this will run in a pod, the kubeconfig will not be actually supplied here, so it should use the secret token mounted within a pod which is called inclusterconfig, an example for that is:
```
 // creates the in-cluster config
    config, err := rest.InClusterConfig()
    if err != nil {
        panic(err.Error())
    }
    // creates the clientset
    clientset, err := kubernetes.NewForConfig(config)
```

File: pkg/upgrade/upgrade.go [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+
+		}
+	}
+	return nil
+}
+
+func PlaceConfigFile(hostPath string, config map[string]interface{}, preservedEntries []string) error {
+	preserved := make(map[string]interface{})
+	if fileExists(hostPath) {
+		logrus.Println("config file exists on the host")
```
Comment (galal-hussein) [inline pkg/upgrade/upgrade.go:124]: can we use Infof instead of println

File: pkg/upgrade/upgrade.go [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+	}
+	if con != "" {
+		selinux.Chcon(fullBinPath, con, true)
+		//selinux.Setfilecon(fullBinPath, filecon)
+	}
+	return nil
+}
+
+func Prepare(masterPlan string) error {
+	kubeconfig := os.Getenv("KUBECONFIG")
```
Comment (galal-hussein) [inline pkg/upgrade/upgrade.go:192]: same here for the in-cluster config

File: pkg/upgrade/upgrade.go (hunk 7 above)
Comment (dweomer) [reply pkg/upgrade/upgrade.go:79]: because the KUBECONFIG will be empty in the pod, the `clientcmd.BuildConfigFromFlags()` call devolves to `rest.InClusterConfig()` to establish the config and clientset from that

File: pkg/upgrade/upgrade.go [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+		return err
+	}
+	var secret types.Secret
+	secretfiles = sortFileByName(secretfiles)
+	for _, secretFile := range secretfiles {
+		fileContent, err := ioutil.ReadFile(secretFile.Name())
+		if err != nil {
+			return err
+		}
+		yaml.Unmarshal(fileContent, secret)
```
Comment (briandowns) [inline pkg/upgrade/upgrade.go:107]: Can you catch and handle the error returned here?

File: pkg/upgrade/upgrade.go [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+func GetBinPath(pid int) (string, error) {
+	_, err := os.FindProcess(pid)
+	if err != nil {
+		return "", fmt.Errorf(" process is not found")
+	}
+	if pid == 1 {
+		return "/bin/k3s", nil
+	}
+	binPath, err := ioutil.ReadFile(path.Join("/host/proc", strconv.Itoa(pid), "cmdline"))
+	if fileExists(string(binPath)) {
```
Comment (dweomer) [inline pkg/upgrade/upgrade.go:63]: I don't know that we need to check that the file exists here. Also, how can it possibly? Won't it be under `/host`?

File: scripts/boilerplate.go.txt [hunk 12]
```diff
@@ -0,0 +1,16 @@
+/*
```
Comment (briandowns) [inline scripts/boilerplate.go.txt:1]: Is this file needed?

File: Dockerfile.dapper [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+RUN mkdir -p /go/src/golang.org/x && \
+    cd /go/src/golang.org/x && git clone https://github.com/golang/tools && \
+    git -C /go/src/golang.org/x/tools checkout -b current aa82965741a9fecd12b026fbb3d3c6ed3231b8f8 && \
+    go install golang.org/x/tools/cmd/goimports
+RUN rm -rf /go/src /go/pkg
+RUN if [ "${ARCH}" == "amd64" ]; then \
+        curl -sL https://install.goreleaser.com/github.com/golangci/golangci-lint.sh | sh -s v1.15.0; \
+    fi
+
+ENV GO111MODULE on
```
Comment (briandowns) [inline Dockerfile.dapper:20]: Is this still needed?

File: pkg/upgrade/upgrade.go [hunk 14]
```diff
@@ -0,0 +1,246 @@
...
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+func fileExists(path string) bool {
```
Comment (dweomer) [inline pkg/upgrade/upgrade.go:30]: This function seems unnecessary. Most if not all of the calls into it should either be skipping the check or making it explicitly on their own (and returning the non "file not found" error conditions to the calling code).

File: pkg/upgrade/upgrade.go [hunk 15]
```diff
@@ -0,0 +1,246 @@
...
+
+func GetBinPath(pid int) (string, error) {
+	_, err := os.FindProcess(pid)
+	if err != nil {
+		return "", fmt.Errorf(" process is not found")
+	}
+	if pid == 1 {
+		return "/bin/k3s", nil
+	}
+	binPath, err := ioutil.ReadFile(path.Join("/host/proc", strconv.Itoa(pid), "cmdline"))
```
Comment (dweomer) [inline pkg/upgrade/upgrade.go:62]: what are we doing with the `err` ?

File: pkg/upgrade/upgrade.go [hunk 16]
```diff
@@ -0,0 +1,246 @@
...
+	if fileExists(string(binPath)) {
+		return string(binPath), nil
+	}
+	return "", fmt.Errorf(" failed to fetch the k3s binary path from process %d", pid)
+}
+
+func TermProcess(pid int) error {
+	proc, err := os.FindProcess(pid)
+	if err != nil {
+		fmt.Errorf(" process is not found")
```
Comment (dweomer) [inline pkg/upgrade/upgrade.go:72]: Should we simply return the `err` here?

File: pkg/upgrade/upgrade.go [hunk 17]
```diff
@@ -0,0 +1,246 @@
...
+}
+
+func MatchSelector(hostPath, configPath string, preservedEntries []string) error {
+	kubeconfig := os.Getenv("KUBECONFIG")
+	conf, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+	if err != nil {
+		return err
+	}
+	nodeClient := kubernetes.NewForConfigOrDie(conf).CoreV1().Nodes()
+	hostname := os.Getenv("SYSTEM_UPGRADE_NODE_NAME")
```
Comment (dweomer) [inline pkg/upgrade/upgrade.go:84]: Prefer to call this `nodeName` instead of `hostname`. This is kind of a nit but it is important to not confuse the two. A kubelet's node name can be different than the hostname it is running on.

File: pkg/upgrade/upgrade.go [hunk 18]
```diff
@@ -0,0 +1,246 @@
...
+	return true
+}
+
+func GetProcessPID() (int, error) {
+	processes, err := ps.Processes()
+	if err != nil {
+		return 0, err
+	}
+	for _, proc := range processes {
+		if strings.HasSuffix(proc.Executable(), "k3s") {
```
Comment (dweomer) [inline pkg/upgrade/upgrade.go:44]: @MonzElmasry when I suggested we use `strings.HasSuffix()` I didn't realize that `proc.Executable` was likely the full `/proc/${pid}/cmdline` which means this won't work as-is, I think.

File: pkg/upgrade/upgrade.go [hunk 19]
```diff
@@ -0,0 +1,246 @@
...
+			return PlaceConfigFile(hostPath, secret.Config, preservedEntries)
+
+		}
+	}
+	return nil
+}
+
+func PlaceConfigFile(hostPath string, config map[string]interface{}, preservedEntries []string) error {
+	preserved := make(map[string]interface{})
+	if fileExists(hostPath) {
```
Comment (dweomer) [inline pkg/upgrade/upgrade.go:123]: Please see my comment above on the `fileExists()` function. I don't think we want to potentially swallow `os.Stat()` errors here, we should return such to callers of `PlaceConfigFile()`

File: pkg/upgrade/upgrade.go [hunk 20]
```diff
@@ -0,0 +1,246 @@
...
+		logrus.Infof("config file exists on the host")
+		filetxt, err := ioutil.ReadFile(hostPath)
```
Comment (dweomer) [inline pkg/upgrade/upgrade.go:125]: This is all the "does the file exist" checking you need to do right here.

File: pkg/upgrade/upgrade.go [hunk 21]
```diff
@@ -0,0 +1,246 @@
...
+	}
+
+	for entry := range preserved {
+		config[entry] = preserved[entry]
+	}
+	content, err := yaml.Marshal(config)
+	if err != nil {
+		return err
+	}
+	return ioutil.WriteFile(hostPath, content, 0755)
```
Comment (dweomer) [inline pkg/upgrade/upgrade.go:148]:
```suggestion
	return ioutil.WriteFile(hostPath, content, 0600)
```

File: pkg/upgrade/upgrade.go (hunk 15 above)
Comment (dweomer) [reply pkg/upgrade/upgrade.go:62]: Additionally, the `/proc/${pid}/cmdline` will contain null (aka zero byte) separated strings IIRC which we will need to deal with before returning.

File: pkg/upgrade/upgrade.go [hunk 22]
```diff
@@ -0,0 +1,246 @@
...
+	return files
+}
+
+func ReplaceBinary() error {
+	newBinaryPath := "/opt/k3s"
+	pid, err := GetProcessPID()
+	if err != nil {
+		return err
+	}
+	fullBinPath, err := GetBinPath(pid)
```
Comment (dweomer) [inline pkg/upgrade/upgrade.go:164]: I am worried that this is returning a string with null bytes in it. See my above comment on `GetBinPath()`

File: pkg/upgrade/upgrade.go [hunk 23]
```diff
@@ -0,0 +1,246 @@
...
+	if err != nil {
+		return err
+	}
+	if !fileExists(newBinaryPath) {
```
Comment (dweomer) [inline pkg/upgrade/upgrade.go:168]: This check is not necessary as any attempt to operate on the specified file path will get you a "file not found" error.

File: pkg/upgrade/upgrade.go [hunk 24]
```diff
@@ -0,0 +1,246 @@
...
+	file, err := filepath.Abs(fullBinPath)
+	if err != nil {
+		return err
+	}
+	con, err := selinux.FileLabel(file)
+	if err != nil {
+		return err
+	}
+	//	con, err := selinux.Getfilecon(fullBinPath)
+	err = os.Rename(fullBinPath, newBinaryPath)
```
Comment (dweomer) [inline pkg/upgrade/upgrade.go:180]: This isn't what we want, I think. We should copy from the `newBinaryPath` to `fullBinPathTmp` where `fullBinPathTmp` is `fullBinPath` plus a process-unique temporary suffix, see https://golang.org/pkg/io/ioutil/#TempFile, and then do `os.Rename(fullBinPathTmp, fullBinPath)`. This achieves two things:
- atomic replacement of the executable (we wouldn't want someone to invoke kubectl while we are in the middle of copying)
- avoids potential `os.Rename()` breakage due to source and target destinations existing under different filesystem mountpoints

File: pkg/upgrade/upgrade.go [hunk 25]
```diff
@@ -0,0 +1,246 @@
...
+	}
+	applyingNodes, found, err := unstructured.NestedSlice(masterplans.Object, "status", "applying")
+	if !found || err != nil {
+		return err
+	}
+	if len(applyingNodes) == 0 {
+		return nil
+	}
+	verifyMasterVersion(os.Getenv("SYSTEM_UPGRADE_PLAN_LATEST_VERSION"))
+	return nil
```
Comment (dweomer) [inline pkg/upgrade/upgrade.go:219]:
```suggestion
	return verifyMasterVersion(os.Getenv("SYSTEM_UPGRADE_PLAN_LATEST_VERSION"))
```
